lib/Reader/readerir.cpp

```diff
@@ -5530,20 +5772,38 @@ IRNode *GenIR::getTypedAddress(IRNode *Addr, CorInfoType CorInfoType,
   // For the 'REFANY' case, verify the address carries
   // reasonable typing. Address producer must ensure this.
   if (CorInfoType == CORINFO_TYPE_REFANY) {
-    PointerType *PointerTy = cast<PointerType>(AddressTy);
-    Type *ReferentTy = PointerTy->getPointerElementType();
-
-    // The result of the load is an object reference,
-    // So addr should be ptr to managed ptr to struct
-    if (!ReferentTy->isPointerTy()) {
-      // If we hit this we should fix the address producer, not
-      // coerce the type here.
-      throw NotYetImplementedException(
-          "unexpected type in load/store primitive");
+    PointerType *PointerTy = dyn_cast<PointerType>(AddressTy);
+    if (PointerTy != nullptr) {
+      Type *ReferentTy = PointerTy->getPointerElementType();
+
+      // The result of the load is an object reference,
+      // So addr should be ptr to managed ptr to struct
+      if (!ReferentTy->isPointerTy()) {
+        // If we hit this we should fix the address producer, not
+        // coerce the type here.
+        throw NotYetImplementedException(
+            "unexpected type in load/store primitive");
+      }
+      PointerType *ReferentPtrTy = cast<PointerType>(ReferentTy);
+      ASSERT(isManagedPointerType(ReferentPtrTy));
+      ASSERT(ReferentTy->getPointerElementType()->isStructTy());
+    } else {
+      // This must be a nativeint, in which case we cast the address
+      // to an address of Object.
+      Type *NativeIntTy =
```

Comment:
The native int represents an unmanaged pointer to a GC pointer, not a GC pointer itself.That shouldn't be a problem for GC tracking, right? In any case, ldind.ref can legally take a nativeint and has to load a GC pointer so I don't see any alternative here.